sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -691,11 +691,11 @@ object DateTimeUtils {
   def truncDate(d: SQLDate, level: Int): SQLDate = {
     def truncToYearLevel(divider: Int, adjust: Int): SQLDate = {
       val oldYear = getYear(d)
-      var newYear = Math.floorDiv(oldYear, divider)
-      if (adjust > 0 && Math.floorMod(oldYear, divider) == 0) {
-        newYear -= 1
+      var newYear = Math.floorDiv(oldYear, divider) * divider
```

Comment:
Sorry for my naive question. Do we need `floorDiv` and `floorMod` here? In my understanding, `oldYear` and `divider` look positive.

Reply:
`oldYear` can be negative, have a look at tests in https://github.com/apache/spark/pull/25230, for instance.

Reply:
I see. The comment `Returns the year value for the given date. The date is expressed in days since 1.1.1970` in `getYear()` means that the value is relative based on 1.1.1970. If the year is older than 1.1.1970, the value is negative.